clients/src/main/java/org/apache/kafka/common/serialization/DoubleDeserializer.java

```diff
@@ -35,4 +41,22 @@ public Double deserialize(String topic, byte[] data) {
         }
         return Double.longBitsToDouble(value);
     }
+
+    @Override
+    public Double deserialize(String topic, Headers headers, ByteBuffer data) {
+        if (data == null) {
+            return null;
+        }
+
+        if (data.remaining() != 8) {
+            throw new SerializationException("Size of data received by DoubleDeserializer is not 8");
+        }
+
+        final ByteOrder srcOrder = data.order();
+        data.order(BIG_ENDIAN);
+
+        final double value = data.getDouble(data.position());
```

Comment:
@LinShunKang , why do we need to set the order to `BIG_ENDIAN` before reading it? From the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#getDouble--):

> Reads the next eight bytes at this buffer's current position, composing them into a double value according to the current byte order, and then increments the position by eight. 

It looks like the byte order is already considered while read from the byte buffer. Had a quick check the jdk source code, and it did check (and convert) the byte order if needed.

Reply:
Also, the `data.position()` could be removed because by default, it'll read from current position of the byte buffer.

Reply:
@showuon 
Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[], and the current byte order of ByteBuffer may not be BIG_ENDIAN, we set the byte order of ByteBuffer to BIG_ENDIAN to be consistent with the byte order used when reading from byte[].

Reply:
> Also, the `data.position()` could be removed because by default, it'll read from current position of the byte buffer.

@showuon 
I use ByteBuffer.getDouble(int) instead of ByteBuffer.getDouble(), considering that ByteBuffer.getDouble(int) does not modify the offsets of the ByteBuffer.

Reply:
@LinShunKang 
> Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[], and the current byte order of ByteBuffer may not be BIG_ENDIAN, we set the byte order of ByteBuffer to BIG_ENDIAN to be consistent with the byte order used when reading from byte[].

Again, from the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#getDouble--):

> Reads the next eight bytes at this buffer's current position, composing them into a double value according to the current byte order, and then increments the position by eight.

The byte order is already considered while reading from the byte buffer. That means, this is not true:
> Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[]

It will read from byte[] using the current byte order. Did I miss anything?

Reply:
> I use ByteBuffer.getDouble(int) instead of ByteBuffer.getDouble(), considering that ByteBuffer.getDouble(int) does not modify the offsets of the ByteBuffer.

I get it. We'll need the position unchanged to get the buffer size (buffer.remaining()). Thanks for the explanation.

Reply:
> @LinShunKang
> 
> > Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[], and the current byte order of ByteBuffer may not be BIG_ENDIAN, we set the byte order of ByteBuffer to BIG_ENDIAN to be consistent with the byte order used when reading from byte[].
> 
> Again, from the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html#getDouble--):
> 
> > Reads the next eight bytes at this buffer's current position, composing them into a double value according to the current byte order, and then increments the position by eight.
> 
> The byte order is already considered while reading from the byte buffer. That means, this is not true:
> 
> > Because DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[]
> 
> It will read from byte[] using the current byte order (BIG_ENDIAN or LITTLE_ENDIAN). Did I miss anything?

In fact, the essence of this issue is which of the following methods we should use for data reading:
1. Force users to write data into ByteBuffer in BIG_ENDIAN mode, and then we read data from ByteBuffer in BIG_ENDIAN mode;
2. We read data according to the byte order of the ByteBuffer passed in by the user by default.

Reply:
> Currently, DoubleDeserializer and other Number Deserializers use BIG_ENDIAN byte order to read from byte[], 

Yes, that's because we parse it by ourselves from byte array. But now, we have API provided by jdk (ByteBuffer) to allow us read according to the byte order in the buffer. So, I'm not convinced why we still need to read using `BIG_ENDIAN` order? Do you have other reason you think should stick to `BIG_ENDIAN` order?

Reply:
> Yes, that's because we parse it by ourselves from byte array. But now, we have API provided by jdk (ByteBuffer) to allow us read according to the byte order in the buffer. So, I'm not convinced why we still need to read using `BIG_ENDIAN` order? Do you have other reason you think should stick to `BIG_ENDIAN` order?

No, that's the only reason I chose method 1.